nixos/modules/services/networking/ucarp.nix

```diff
@@ -0,0 +1,166 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.networking.ucarp;
+
+  upscript = pkgs.writeScript "upscript" cfg.upscript;
+  downscript = pkgs.writeScript "downscript" cfg.downscript;
+
+  ucarpExec = concatStringsSep " " (
+    [
+      "${pkgs.ucarp}/bin/ucarp"
+      "--interface=${cfg.interface}"
+      "--srcip=${cfg.srcIp}"
+      "--vhid=${toString cfg.vhId}"
+      "--pass=${cfg.pass}"
+      "--addr=${cfg.addr}"
+      "--advbase=${toString cfg.advBase}"
+      "--advskew=${toString cfg.advSkew}"
+      "--upscript=${upscript}"
+      "--downscript=${downscript}"
+      "--deadratio=${toString cfg.deadratio}"
+    ]
+    ++ (optional cfg.preempt "--preempt")
+    ++ (optional cfg.neutral "--neutral")
+    ++ (optional cfg.shutdown "--shutdown")
+    ++ (optional cfg.ignoreIfState "--ignoreifstate")
+    ++ (optional cfg.noMcast "--nomcast")
+    ++ (optional (cfg.extraParam != null) "--xparam=${cfg.extraParam}")
+  );
+in {
+  options.networking.ucarp = {
+    enable = mkEnableOption "ucarp, userspace implementation of CARP";
+
+    interface = mkOption {
+      type = types.str;
+      description = "Network interface to bind to";
+      example = "eth0";
+    };
+
+    srcIp = mkOption {
+      type = types.str;
+      description = "Source (real) IP address of this host";
+    };
+
+    vhId = mkOption {
+      type = types.ints.between 1 255;
+      description = "Virtual IP identifier (1-255)";
+      example = 1;
+    };
+
+    pass = mkOption {
+      type = types.str;
+      description = "Shared password";
+    };
+
+    preempt = mkOption {
+      type = types.bool;
+      description = "Preemptive failover, becomes a master as soon as possible";
+      default = false;
+    };
+
+    neutral = mkOption {
+      type = types.bool;
+      description = "Do not run downscript at start if backup";
+      default = false;
+    };
+
+    addr = mkOption {
+      type = types.str;
+      description = "Virtual shared IP address";
+    };
+
+    advBase = mkOption {
+      type = types.ints.unsigned;
+      description = "Advertisement frequency in seconds";
+      default = 1;
+    };
+
+    advSkew = mkOption {
+      type = types.ints.unsigned;
+      description = "Advertisement skew in seconds";
+      default = 0;
+    };
+
+    upscript = mkOption {
```

Comment:
You might consider making this `type` a `path` and then using `writeShellScript` (or something like that) in the `example`. Even if you don't drink enough nix kool-aid to buy into "`/bin/sh` is impure and impure is bad" argument this module is going to be part of nixpkgs and should probably promote these ideals.

Reply:
Done.